presto-cassandra/src/main/java/com/facebook/presto/cassandra/NativeCassandraSession.java

```diff
@@ -406,20 +417,59 @@ public ResultSet execute(Statement statement)
         return executeWithSession(session -> session.execute(statement));
     }
 
-    private Iterable<Row> queryPartitionKeys(CassandraTable table, List<Object> filterPrefix)
+    private Iterable<Row> queryPartitionKeys(CassandraTable table, List<Set<Object>> filterPrefixes)
     {
         CassandraTableHandle tableHandle = table.getTableHandle();
         List<CassandraColumnHandle> partitionKeyColumns = table.getPartitionKeyColumns();
 
-        if (filterPrefix.size() != partitionKeyColumns.size()) {
+        if (filterPrefixes.size() != partitionKeyColumns.size()) {
             return null;
         }
 
         Select partitionKeys = CassandraCqlUtils.selectDistinctFrom(tableHandle, partitionKeyColumns);
-        addWhereClause(partitionKeys.where(), partitionKeyColumns, filterPrefix);
+        addWhereInClauses(partitionKeys.where(), partitionKeyColumns, filterPrefixes);
+
         return execute(partitionKeys).all();
     }
 
+    private Iterable<Row> queryPartitionKeysLegacy(CassandraTable table, List<Set<Object>> filterPrefixes)
+    {
+        CassandraTableHandle tableHandle = table.getTableHandle();
+        List<CassandraColumnHandle> partitionKeyColumns = table.getPartitionKeyColumns();
+
+        if (filterPrefixes.size() != partitionKeyColumns.size()) {
+            return null;
+        }
```

Comment:
```
+        CassandraTableHandle tableHandle = table.getTableHandle();
 +        List<CassandraColumnHandle> partitionKeyColumns = table.getPartitionKeyColumns();
 +
 +        if (filterPrefixes.size() != partitionKeyColumns.size()) {
 +            return null;
 +        }
```
above looks to be common with `queryPartitionKeys`, maybe it should be executed before this method (in `queryPartitionKeys`)?